sys/net/application_layer/nanocoap/nanocoap.c

```diff
@@ -614,20 +614,6 @@ size_t coap_put_option(uint8_t *buf, uint16_t lastonum, uint16_t onum, const uin
     return (size_t)n;
 }
 
-size_t coap_put_option_ct(uint8_t *buf, uint16_t lastonum, uint16_t content_type)
-{
-    if (content_type == 0) {
-        return coap_put_option(buf, lastonum, COAP_OPT_CONTENT_FORMAT, NULL, 0);
-    }
-    else if (content_type <= 255) {
-        uint8_t tmp = content_type;
-        return coap_put_option(buf, lastonum, COAP_OPT_CONTENT_FORMAT, &tmp, sizeof(tmp));
-    }
-    else {
-        return coap_put_option(buf, lastonum, COAP_OPT_CONTENT_FORMAT, (uint8_t *)&content_type, sizeof(content_type));
-    }
-}
-
 static unsigned _size2szx(size_t size)
```

Comment:
Sine this PR aims at optimizing, does it make sense using `_size2szx()` every time? Would it make sense refactoring and having data types for the blocksizes:

```
typedef enum {
    COAP_BLOCKSIZE_32 = 1,
    COAP_BLOCKSIZE_64,
    COAP_BLOCKSIZE_128,
    COAP_BLOCKSIZE_256,
    COAP_BLOCKSIZE_512,
    COAP_BLOCKSIZE_1024,
} coap_blksize_t;
```

It would avoid doing the loop every time.

Reply:
Thanks for the suggestion, @fjmolinas. I don't follow how to refactor use of _size2szx() to include the enum though. I would appreciate more detail.

Reply:
I think the idea was to not allow arbitrary sizes to be passed to be passed around, but only `coap_blksize_t` that don't need to be converted again.

So you'd have `coap_blksize_t blksize` instead of `size_t blksize`.

Reply:
> I think the idea was to not allow arbitrary sizes to be passed to be passed around, but only coap_blksize_t that doesn't need to be converted again.
> 
> So you'd have coap_blksize_t blksize instead of size_t blksize.

Yes, that was my idea. But looking at the code again I guess doing this would mean changing the `slicer` a bit so it performs a shift on the received size. I should have though of this when reviewing #11024, but I think it is a bit out of scope for this PR, I leave it up to you @kb2ma :)

In any case the while loop could be avoided by using [`__builtin_clz`](http://www.keil.com/support/man/docs/armcc/armcc_chr1359124994619.htm) here: https://github.com/RIOT-OS/RIOT/blob/c1aa872c11f66788efa994394bbd3613e45ec5e0/sys/net/application_layer/nanocoap/nanocoap.c#L661-L664

Reply:
Yes, use of the enum would require modification of the slicer. I will defer any changes there. Eventually, I would like to see a "payload writer" that uses blockwise by default. That would be a good time to revisit the capabilities of the slicer.

Thanks for the reference to the builtin functions. I have not used them before. It looks like  bitarithm_lsb() wraps the use of the clz feature based on availability, so that's what I went with. Let me know if there is a better way.

Reply:
Looks good!